Record delisted pools in a dedicated table.

jonathanknowles opened the pull request:
# Issue Number

#2249 

# Overview

This PR records delisted pools in a dedicated table instead of using a field in the `pool_registrations` table.

In the updated schema, a pool is delisted if (and only if) there is a single row containing that pool's id in the `delisted_pools` table.

This solution has several advantages:

1.  We only need a single database row to record that a pool is delisted.

2.  We no longer need to carefully to ensure that all registration records for a particular pool have the same delisted status. A pool is either _delisted_ or _not delisted_: the schema rules out all intermediate states.

3.  Pools automatically remain delisted when rollbacks occur or when new certificates are published, with no special effort.

4.  The `putPoolRegistration` function no longer needs to read the  most-recently-written registration entry before adding a new entry.

5.  Each row in the `pool_registrations` table is now just an immutable record of a registration certificate.

jonathanknowles committed:
Record delisted pools in a dedicated table.

Record delisted pools in a dedicated table instead of using a field in
the `pool_registrations` table.

In the updated schema, a pool is delisted if (and only if) there is a
single row containing that pool's id in the `delisted_pools` table.

This solution has several advantages:

  1.  We only need a single database row to record that a pool is delisted.

  2.  We no longer need to carefully to ensure that all registration records
      for a particular pool have the same delisted status. A pool is either
      delisted or not delisted: the schema rules out all intermediate states.

  3.  Pools automatically remain delisted when rollbacks occur or when new
      certificates are published, with no extra effort.

  4.  The `putPoolRegistration` function no longer needs to read the
      most-recently-written registration entry before adding a new entry.

  5.  Each row in the `pool_registrations` table is now just an immutable
      record of a registration certificate.

  6.  The `PoolFlag` type is no longer necessary.

hasufell commented on lib/core/src/Cardano/Pool/DB/Model.hs:
```diff
@@ -152,6 +154,8 @@ data PoolDatabase = PoolDatabase
         !(Map (CertificatePublicationTime, PoolId) PoolRetirementCertificate)
     -- ^ On-chain retirements associated with pools
 
+    , delisted :: !(Set PoolId)
```
I wonder if we should just enable `StrictData`, but that probably warrants a discussion.

jonathanknowles replied:
We could. Though I agree we'd have to make it a wider discussion.

hasufell commented on lib/core/src/Cardano/Pool/DB/Model.hs:
```diff
@@ -430,16 +432,10 @@ mRollbackTo ti point = do
-  where
-    updateThis p = p `Set.member` poolsToDelistSet
-    poolsToDelistSet = Set.fromList poolsToDelist
+mDelistPools = modify #delisted . Set.union . Set.fromList
```
Ok, so I think we don't want union here! The SMASH server has an endpoint to enlist already delisted pools afais:

https://github.com/input-output-hk/smash/blob/f4a179a1e2de22d332ec731df8fedaed5d7d8754/src/Cardano/SMASH/API.hs#L60

That means we really want to overwrite here, not merge.

hasufell replied:
And we should also test this property. `delistPools` overwrites, it doesn't add.

jonathanknowles replied:
Aha, I didn't realize that `delistPools` was supposed to completely replace the set each time.

In that case, I think the `delistPools` name is a bit confusing, as it sounds as though it ought to augment the existing set, rather than replacing it.

Perhaps call the functions `putDelistedPools` and `readDelistedPools`?

jonathanknowles replied:
> That means we really want to overwrite here, not merge.

Fixed in b67c89fd8aed7b2349dd18ac4a0f8ed9e981823b.

jonathanknowles replied:
> And we should also test this property. delistPools overwrites, it doesn't add.

Fixed in b3f3d7ed45cde2a603b9fe26a6a916c89fd97132.

hasufell commented on lib/core/src/Cardano/Pool/DB/Sqlite.hs:
```diff
@@ -500,13 +491,11 @@ newDBLayer trace fp timeInterpreter = do
-                (\(Entity k val) -> (k, val {poolRegistrationFlag = Delisted}))
-                px
+        delistPools =
+            insertMany_ . fmap DelistedPool
```
Same as above. We need to drop the table beforehand.

jonathanknowles replied:
> Same as above. We need to drop the table beforehand.

Fixed in b67c89fd8aed7b2349dd18ac4a0f8ed9e981823b.

hasufell commented on lib/core/src/Cardano/Pool/DB/Sqlite/TH.hs:
```diff
@@ -120,11 +120,15 @@ PoolRegistration sql=pool_registration
     Primary poolRegistrationPoolId poolRegistrationSlot poolRegistrationSlotInternalIndex
     deriving Show Generic
 
+DelistedPool sql=delisted_pool
```
I'd propose to call the table `pool_delisted` or `pool_delistment` (is that even a word?). The other tables also start with `pool_`.

hasufell replied:
That might also fix the naming issue below

jonathanknowles replied:
Fixed in 35e29f854a2769a659f75903b56bc3ad4a2c7146.

hasufell commented on lib/core/src/Cardano/Wallet/Primitive/Types.hs:
```diff
@@ -2042,14 +2039,3 @@ instance FromJSON PoolMetadataSource where
 instance ToJSON PoolMetadataSource where
     toJSON = toJSON . toText
-
-data PoolFlag = NoPoolFlag | Delisted
```
One naming discussion less :)

hasufell commented on lib/core/test/unit/Cardano/Pool/DB/Properties.hs:
```diff
@@ -1513,16 +1511,15 @@ prop_delistPoolsPersists DBLayer {..} cert =
         let poolid = view #poolId . snd $ cert
         -- delist pool
         run $ atomically $ delistPools [poolid]
-        delisted <- run . atomically . readPoolRegistration $ poolid
```
This whole test doesn't make sense anymore. Because the delisted flag doesn't exist in the table in question. So it doesn't test anything.

jonathanknowles replied:
I agree that we could delete this test.

Though given your [earlier comment](https://github.com/input-output-hk/cardano-wallet/pull/2277#discussion_r513384085) about completely overwriting the set each time we do a `delistPools`, I think it would make sense to update the test so that it checks that the set is replaced, rather than augmented.

hasufell replied:
Sure, but that's a different property. This test was about "persistence in light of certificate insertion".

jonathanknowles replied:
> Sure, but that's a different property. This test was about "persistence in light of certificate insertion".

I've now deleted this test.

Fixed in c94d57890fd9ce1099e06731a0e66f2bce6bde63.

hasufell commented on lib/shelley/src/Cardano/Wallet/Shelley/Pools.hs:
```diff
@@ -560,6 +562,7 @@ readPoolDbData DBLayer {..} currentEpoch = atomically $ do
         retirementCertificates
         <$> readTotalProduction
         <*> readPoolMetadata
+        <*> (Set.fromList <$> readDelistedPools)
```
Maybe we just pass it through as a Set? There's json instances for it anyway.

jonathanknowles replied:
By passing it through as a `Set`, do you mean writing a set to a single row?

I think that would make it harder to inspect the table using tools like `sqlitebrowser`.

In any case, having one row per pool ID means that we can more easily write queries on the table, especially if we need to perform a join elsewhere.

hasufell replied:
Nvm... it's fine this way

hasufell reviewed (changes requested):
Looks good, just a few remarks

jonathanknowles committed:
Rename `delisted_pools` table to `pool_delistment`.

In response to review feedback:

https://github.com/input-output-hk/cardano-wallet/pull/2277#discussion_r513385694

jonathanknowles committed:
Make `delistPools` replace the set of delisted pools.

We must replace the set of delisted pools, rather than augmenting it.

In response to review feedback:

https://github.com/input-output-hk/cardano-wallet/pull/2277#discussion_r513384085

jonathanknowles committed:
Rename `delistPools` to `putDelistedPools`.

This operation completely replaces the set of delisted pools, rather
than augmenting it. Therefore, the name `delistPools` is slightly
misleading, as it gives the impression that the existing set will be
augmented, which is no longer true.

In response to review feedback:

https://github.com/input-output-hk/cardano-wallet/pull/2277#discussion_r513384085

jonathanknowles committed:
Update `prop_putDelistedPools` to check for overwriting.

In particular, we check that 'putDelistedPools' completely overwrites
the existing set every time.

In response to review feedback:

https://github.com/input-output-hk/cardano-wallet/pull/2277#discussion_r513390601

jonathanknowles committed:
Remove `prop_putDelistedPoolsPersists`.

This test no longer makes sense, as delisted pools are no longer stored
in the `pool_registration` table.

In response to review feedback:

https://github.com/input-output-hk/cardano-wallet/pull/2277#discussion_r513387397

hasufell commented on lib/core/src/Cardano/Pool/DB/Model.hs:
```diff
@@ -453,6 +449,8 @@ mRemovePools poolsToRemove = do
         $ Map.filterWithKey $ \(_, p) _ -> retain p
     modify #retirements
         $ Map.filterWithKey $ \(_, p) _ -> retain p
+    modify #delisted
```
Do we really want this? I'd say that the pool delisted set should just reflect what SMASH returns, without any local modifications?

hasufell replied:
Also note: the next query to SMASH will add the just deleted poolIds again.

jonathanknowles replied:
Well, it all depends on whether you want the set of pool IDs in `pool_delistment` to always be a subset of the pools that we know about.

I'd say it's probably okay to remove them when we remove pools.

hasufell replied:
But we're not yet satisfying that property, see

https://github.com/input-output-hk/cardano-wallet/pull/2277/files#diff-7221813f41d4c796791728852d1b5e3f2007929b3c613e3d07dda5484d11d486R494-R496

jonathanknowles replied:
> Do we really want this? I'd say that the pool delisted set should just reflect what SMASH returns, without any local modifications?

Fair enough. I've fixed this in 12f5426917b675d1c9ee01fc2f48510d95ecf422.

hasufell commented on lib/core/src/Cardano/Pool/DB/Sqlite.hs:
```diff
@@ -515,6 +505,7 @@ newDBLayer trace fp timeInterpreter = do
             deleteWhere [ PoolRegistrationPoolId ==. pool ]
             deleteWhere [ PoolRetirementPoolId ==. pool ]
             deleteWhere [ StakeDistributionPoolId ==. pool ]
+            deleteWhere [ DelistedPoolId ==. pool ]
```
Same as above.

jonathanknowles replied:
Fixed in 12f5426917b675d1c9ee01fc2f48510d95ecf422.

jonathanknowles committed:
Adjust `removePools` to not remove pools from the delisted set.

In response to review feedback:

https://github.com/input-output-hk/cardano-wallet/pull/2277#discussion_r513436734
